### PR TITLE
Use rules to add computers as virtual machines

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes/no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Review by orthagh — APPROVED** [review]

File: src/Inventory/Asset/VirtualMachine.php [hunk 1]
```diff
@@ -302,36 +301,53 @@ protected function createVmComputer()
...
                 if ($computers_vm_id == 0) {
-                   // Add computer
-                    $vm->entities_id = $this->item->fields['entities_id'];
-                    $computers_vm_id = $computervm->add(Toolbox::addslashes_deep((array)$vm));
+                    //call rules on current collected data to find item
+                    //a callback on rulepassed() will be done if one is found.
+                    $rule = new RuleImportAssetCollection();
+                    $rule->getCollectionPart();
+                    $input = (array)$vm;
+                    $input ['itemtype'] = \Computer::class;
```
Comment (cedric-anne) [inline src/Inventory/Asset/VirtualMachine.php:309]:
```suggestion
                    $input['itemtype'] = \Computer::class;
```

File: src/Inventory/Asset/VirtualMachine.php [hunk 2]
```diff
@@ -302,36 +301,53 @@ protected function createVmComputer()
...
+                    $input['entities_id'] = $this->main_asset->getEntityID();
+                    $input  = \Toolbox::addslashes_deep($input);
+                    $datarules = $rule->processAllRules($input);
+
+                    if (isset($datarules['_no_rule_matches']) and ($datarules['_no_rule_matches'] == '1') or isset($datarules['found_inventories'])) {
```
Comment (cedric-anne) [inline src/Inventory/Asset/VirtualMachine.php:314]:
```suggestion
                    if (isset($datarules['_no_rule_matches']) && ($datarules['_no_rule_matches'] == '1') || isset($datarules['found_inventories'])) {
```

File: src/Inventory/Asset/VirtualMachine.php [hunk 3]
```diff
@@ -302,36 +301,53 @@ protected function createVmComputer()
...
+                        //this is a new one
+                        $vm->entities_id = $this->item->fields['entities_id'];
+                        $computers_vm_id = $computervm->add($input);
+                    } else {
+                        //refused by rules
+                        return;
+                    }
+                    // Add computer
```
Comment (cedric-anne) [inline src/Inventory/Asset/VirtualMachine.php:322]: Should this comment be placed somewhere else ?

Comment (trasher) [inline src/Inventory/Asset/VirtualMachine.php:322]:
```suggestion
```

Comment (trasher) [reply src/Inventory/Asset/VirtualMachine.php]: (I've removed this one)

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**